internal/v1/handler.go

```diff
@@ -223,35 +227,114 @@ func (h *Handlers) GetPackages(ctx echo.Context, params GetPackagesParams) error
 	})
 }
 
+func shouldRetry(err error) bool {
+	if e, ok := err.(net.Error); ok {
+		// "e.Temporary" is deprecated so let's use e.Timeout
+		// as an approximation. We may need to handle ECONNRESET
+		// or similar but let's do as we go
+		// https://github.com/golang/go/issues/45729
+		return e.Timeout()
+	}
+	return false
+}
+
+// getBodyWithRetry will run the given function and return the reponse
+// body. On error the function is retried when:
+// - the error is a timeout
+// - the error is a retryable http code (e.g. 503)
+// Note that even on error the body can be non-nil, in this case it will
+// contain the body of the errored http response.
+func (h *Handlers) getBodyWithRetry(ctx echo.Context, f func() (*http.Response, error), info string) ([]byte, error) {
+	expectedHttpStatus := http.StatusOK
+	return h.getBodyWithRetryInternal(ctx, expectedHttpStatus, f, info)
+}
+
+// postBodyWithRetry is similar to getBodyWithRetry() but it expects a
+// http.StatusCreated. Any other return code is an error.
+func (h *Handlers) postBodyWithRetry(ctx echo.Context, f func() (*http.Response, error), info string) ([]byte, error) {
+	expectedHttpStatus := http.StatusCreated
+	return h.getBodyWithRetryInternal(ctx, expectedHttpStatus, f, info)
+}
+
+func (h *Handlers) getBodyWithRetryInternal(ctx echo.Context, expectedHttpStatus int, f func() (*http.Response, error), info string) ([]byte, error) {
+	// The list of http errors we want to retry on. be convervative here
+	// at first to avoid unneccessary retries for unrecoverable err and
+	// also the thundering herd problem
+	retryableHttpCodes := []int{http.StatusServiceUnavailable}
+
+	// XXX: some tests make decisions based on the http reponse body so
+	// we include it here even though it's a weird interface to return
+	// both data and an error - we should rethink this but it needs more
+	// refactoring
+	var body []byte
+	nRetries := 0
+	err := retry.Do(
+		func() error {
+			nRetries++
+			resp, err := f()
+			if err != nil {
+				if shouldRetry(err) {
+					return err
+				}
+				return retry.Unrecoverable(echo.NewHTTPError(http.StatusInternalServerError, fmt.Sprintf("Failed %v: %v (attempts: %v)", info, err, nRetries)))
+			}
+			defer closeBody(ctx, resp.Body)
+
+			body, err = io.ReadAll(resp.Body)
```

Comment:
There is maybe a subtle bug here - if we do a POST but fail to read the reply from composer we will retry - is this safe, i.e. will we possibly create multiple jobs just because composer could not sent us all the data (but created the job)? Probably needs checking on the composer side as well.